File: dotnet/Allmark/Block/ListRule.cs
namespace Allmark.Block;

using Allmark.Types;

public record ListInfo
{
	public required string Delimiter { get; set; }
	public required string Markup { get; set; }
	public required bool IsBlank { get; set; }
	public required string Type { get; set; }
}

public static class ListRule
{
	public static bool TestListStart(
		BlockParserState state,
		MarkdownNode parent,
		ListInfo? info)
	{
		if (info == null)
		{
			return false;
		}

		MarkdownNode? closedNode = null;

		// TODO: Move the things that refer to "list_[un]ordered" into their rules

		// "When the first list item in a list interrupts a paragraph—that is, when
		// it starts on a line that would otherwise count as paragraph continuation
		// text—then (a) the lines Ls must not begin with a blank line, and (b) if
		// the list item is ordered, the start number must be 1."
		// HACK: Only at the top level??
		if (parent.Type == "paragraph" && state.OpenNodes.Count == 2)
		{
			// "[A]n empty list item cannot interrupt a paragraph"
			if (info.IsBlank)
			{
				return false;
			}

			// "[W]e allow only lists starting with 1 to interrupt paragraphs"
			if (info.Type == "list_ordered" && info.Markup != "1" + info.Delimiter)
			{
				return false;
			}
		}

		// "[L]ist items may not be indented more than three spaces"
		var openIndent = state.Indent;
		for (var i = 0; i < state.OpenNodes.Count - 1; i++)
		{
			if (IsList(state.OpenNodes.ElementAt(i).Type))
			{
				openIndent -= state.OpenNodes.ElementAt(i).Indent;
				break;
			}
		}
		if (openIndent >= 4)
		{
			return false;
		}

		// If this was possibly a continuation, it no longer is
		// If there was a blank line after the paragraph, move it to the continued node
		if (state.MaybeContinue)
		{
			state.MaybeContinue = false;
			for (var j = 0; j < state.OpenNodes.Count - 1; j++)
			{
				var node = state.OpenNodes.ElementAt(j);
				if (node.MaybeContinuing)
				{
					node.MaybeContinuing = false;
					closedNode = node;
					// Pop until we reach this node
					var newLength = state.OpenNodes.Count - j - 1;
					while (state.OpenNodes.Count > newLength)
					{
						state.OpenNodes.Pop();
					}
				}
			}
			parent = state.OpenNodes.Peek();
		}

		// If there's an open paragraph, close it
		// TODO: Is there a better way to do this??
		if (parent.Type == "paragraph")
		{
			closedNode = state.OpenNodes.Pop();
			parent = state.OpenNodes.Peek();
		}

		// If there's an open list of a different type, and this node is not nested, close it
		// TODO: Is there a better way to do this??
		if (IsList(parent.Type) && parent.Delimiter != info.Delimiter)
		{
			var lastItem = parent.Children!.Last();
			if (lastItem?.Type == "list_item" && state.Indent < lastItem.Subindent)
			{
				closedNode = state.OpenNodes.Pop();
				parent = state.OpenNodes.Peek();
			}
		}

		if (closedNode != null)
		{
			Utils.CloseNode(state, closedNode);
		}

		// Count spaces after the marker
		var spaces = 0;
		var blank = true;
		for (var j = state.I + info.Markup.Length; j < state.Src.Length; j++)
		{
			var nextChar = state.Src[j];
			if (Utils.IsNewLine(nextChar))
			{
				break;
			}
			else if (Utils.IsSpace(nextChar))
			{
				spaces++;
			}
			else
			{
				blank = false;
				break;
			}
		}

		// If there's a newline after the marker, treat it as one space
		if (blank)
		{
			spaces = 1;
		}

		// "If the first block in the list item is an indented code block, then
		// by rule #2, the contents must be indented one space after the list
		// marker:"
		if (spaces > 4)
		{
			spaces = 1;
		}

		var haveList = parent.Type == info.Type;
		var list = haveList
			? parent
			: Utils.NewNode(info.Type, true, state.I, state.Line, 1, info.Markup, state.Indent, []);
		list.Delimiter = info.Delimiter;
		var item = Utils.NewNode("list_item", true, state.I, state.Line, 1, info.Markup, state.Indent, []);
		item.Delimiter = info.Delimiter;
		item.Subindent = state.Indent + info.Markup.Length + spaces;

		if (!haveList)
		{
			if (state.HasBlankLine && parent.Children!.Count > 0)
			{
				parent.Children[^1].BlankAfter = true;
				state.HasBlankLine = false;
			}

			parent.Children!.Add(list);
			state.OpenNodes.Push(list);
		}

		if (state.HasBlankLine && parent.Children!.Count > 0)
		{
			parent.Children[^1].BlankAfter = true;
			state.HasBlankLine = false;
		}

		list.Children!.Add(item);
		state.OpenNodes.Push(item);

		Utils.MovePastMarker(info.Markup.Length, state);

		state.HasBlankLine = false;
		Parse.ParseBlock.Execute(state, item);

		return true;
	}

	public static bool TestListContinue(
		BlockParserState state,
		MarkdownNode node,
		ListInfo? info)
	{
		var ch = Utils.GetChar(state.Src, state.I);

		// If there's the same list marker and the indent is not too far, we can continue
		if (info != null)
		{
			if (state.HasBlankLine && state.Indent >= 4)
			{
				return false;
			}
			if (info.Delimiter == node.Delimiter)
			{
				return true;
			}
		}

		// Can't continue if there's only one item, it's blank and there's a blank line after the list
		// HACK: This is messy
		if (state.HasBlankLine && node.Children!.Count == 1 && node.Children[0].Children!.Count == 0)
		{
			return false;
		}

		// TODO: Not sure about this one
		// Also, do the state.isEmpty check with indent like on the other branch
		if (Utils.IsNewLine(ch))
		{
			return true;
		}

		var openNode = state.OpenNodes.Peek();
		if (openNode.Type == "paragraph")
		{
			// We won't know until we try more things
			return true;
		}

		var lastItem = node.Children!.Last();
		if (lastItem != null && lastItem.Type == "list_item" && state.Indent >= lastItem.Subindent)
		{
			return true;
		}

		return false;
	}

	// HACK: Not great
	private static bool IsList(string nodeType)
	{
		return nodeType.StartsWith("list_") && nodeType != "list_item";
	}
}

File: zig/src/block/listRule.zig
const std = @import("std");
const BlockParserState = @import("../types/BlockParserState.zig").BlockParserState;
const BlockRule = @import("../types/BlockRule.zig").BlockRule;
const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const parseBlock = @import("../parse/parseBlock.zig").parseBlock;
const isNumeric = @import("../utils/isAlphaNumeric.zig").isNumeric;
const isNewLine = @import("../utils/isNewLine.zig").isNewLine;
const isSpace = @import("../utils/isSpace.zig").isSpace;
const movePastMarker = @import("../utils/movePastMarker.zig").movePastMarker;
const newBlock = @import("../utils/newBlock.zig").newBlock;
const closeNode = @import("../utils/closeNode.zig").closeNode;
const appendChild = @import("../utils/appendChild.zig").appendChild;

pub fn getMarkup(state: *BlockParserState) ?ListInfo {
    if (state.i >= state.src.len) return null;
    const char = state.src[state.i];

    if (char != '-' and char != '+' and char != '*') return null;

    const is_bullet = (state.i + 1 >= state.src.len) or isSpace(state.src[state.i + 1]);

    if (!is_bullet) return null;

    var next_char: u8 = 0;
    if (state.i + 1 < state.src.len) {
        next_char = state.src[state.i + 1];
    }

    return ListInfo{
        .delimiter = char,
        .markup = state.src[state.i .. state.i + 1],
        .is_blank = (state.i + 1 >= state.src.len) or next_char == '\n' or next_char == '\r',
        .type = "list_bulleted",
    };
}

pub fn testListStart(state: *BlockParserState, parent: *MarkdownNode, end_of_line: usize, info: ?ListInfo) bool {
    if (info == null) {
        return false;
    }

    return testStartWithInfo(state, parent, end_of_line, info.?);
}

fn testStartWithInfo(state: *BlockParserState, parent: *MarkdownNode, end_of_line: usize, info: ListInfo) bool {
    var effective_parent = parent;
    var closed_node: ?*MarkdownNode = null;

    if (std.mem.eql(u8, parent.type, "paragraph") and state.openNodes.items.len == 2) {
        const info_blank = info.is_blank;
        if (info_blank) return false;

        const is_ordered_not_first = std.mem.eql(u8, info.type, "list_ordered") and !(info.markup.len == 2 and info.markup[0] == '1');
        if (is_ordered_not_first) return false;
    }

    var open_indent: i32 = state.indent;
    var i: usize = state.openNodes.items.len;
    while (i > 1) {
        i -= 1;
        const open_node = state.openNodes.items[i];
        const is_list_parent = std.mem.eql(u8, open_node.type, "list_bulleted") or std.mem.eql(u8, open_node.type, "list_ordered");
        if (is_list_parent) {
            open_indent -= open_node.indent;
            break;
        }
    }

    if (open_indent >= 4) {
        return false;
    }

    if (state.maybeContinue) {
        state.maybeContinue = false;
        var j: usize = state.openNodes.items.len;
        while (j > 1) {
            j -= 1;
            const node = state.openNodes.items[j];
            if (node.maybeContinuing) {
                node.maybeContinuing = false;
                closed_node = node;
                state.openNodes.shrinkRetainingCapacity(j);
            }
        }
        effective_parent = state.openNodes.items[state.openNodes.items.len - 1];
    }

    // If there's an open paragraph, close it
    if (std.mem.eql(u8, effective_parent.type, "paragraph")) {
        closed_node = state.openNodes.pop();
        effective_parent = state.openNodes.items[state.openNodes.items.len - 1];
    }

    const is_list_parent = std.mem.eql(u8, effective_parent.type, "list_bulleted") or std.mem.eql(u8, effective_parent.type, "list_ordered");
    const is_diff_type = is_list_parent and (effective_parent.delimiter.len == 0 or effective_parent.delimiter[0] != info.delimiter);

    if (is_diff_type) {
        const last_item: ?*MarkdownNode = if (effective_parent.children) |children| blk: {
            if (children.len > 0) break :blk children[children.len - 1] else break :blk null;
        } else null;

        if (last_item) |li| {
            if (std.mem.eql(u8, li.type, "list_item") and state.indent < li.subindent) {
                closed_node = state.openNodes.pop();
                effective_parent = state.openNodes.items[state.openNodes.items.len - 1];
            }
        }
    }

    if (closed_node) |cn| {
        closeNode(state, cn);
    }

    var spaces: usize = 0;
    var blank: bool = true;
    var j: usize = state.i + info.markup.len;
    while (j < state.src.len) : (j += 1) {
        if (state.src[j] == '\n' or state.src[j] == '\r') break;
        if (isSpace(state.src[j])) {
            spaces += 1;
        } else {
            blank = false;
            break;
        }
    }

    if (blank) {
        spaces = 1;
    }

    // "If the first block in the list item is an indented code block, then
    // by rule #2, the contents must be indented one space after the list
    // marker:"
    if (spaces > 4) {
        spaces = 1;
    }

    const has_list = std.mem.eql(u8, effective_parent.type, info.type);
    const list = if (has_list) effective_parent else newBlock(state.allocator, info.type, state.i, state.line, info.markup, state.indent) catch unreachable;
    const item = newBlock(state.allocator, "list_item", state.i, state.line, info.markup, state.indent) catch unreachable;

    if (!has_list) {
        const delimiter_char = &[_]u8{info.delimiter};
        list.delimiter = state.allocator.dupe(u8, delimiter_char) catch unreachable;
        list.delimiter_allocated = true;
    }

    const item_delimiter_char = &[_]u8{info.delimiter};
    item.delimiter = state.allocator.dupe(u8, item_delimiter_char) catch unreachable;
    item.delimiter_allocated = true;
    item.subindent = state.indent + @as(i32, @intCast(info.markup.len)) + @as(i32, @intCast(spaces));

    if (!has_list) {
        if (state.hasBlankLine and effective_parent.children != null and effective_parent.children.?.len > 0) {
            const last_child = effective_parent.children.?[effective_parent.children.?.len - 1];
            last_child.blankAfter = true;
            state.hasBlankLine = false;
        }
        appendChild(state.allocator, effective_parent, list) catch unreachable;
        state.openNodes.append(state.allocator, list) catch unreachable;
    }

    if (state.hasBlankLine and effective_parent.children != null and effective_parent.children.?.len > 0) {
        const last_child = effective_parent.children.?[effective_parent.children.?.len - 1];
        last_child.blankAfter = true;
        state.hasBlankLine = false;
    }

    if (list.children == null) {
        list.children = state.allocator.alloc(*MarkdownNode, 0) catch unreachable;
    }
    item.children = state.allocator.alloc(*MarkdownNode, 0) catch unreachable;
    appendChild(state.allocator, list, item) catch unreachable;
    state.openNodes.append(state.allocator, item) catch unreachable;

    movePastMarker(info.markup.len, state);
    state.hasBlankLine = false;
    parseBlock(state, item, end_of_line);

    return true;
}

pub fn testListContinue(state: *BlockParserState, node: *MarkdownNode, info: ?ListInfo) bool {
    return testContinue(info, state, node);
}

pub fn testContinue(info: ?ListInfo, state: *BlockParserState, node: *MarkdownNode) bool {
    if (state.i >= state.src.len) return false;
    const char = state.src[state.i];

    if (info != null) {
        if (state.hasBlankLine and state.indent >= 4) {
            return false;
        }
        if (info.?.delimiter == node.delimiter[0]) {
            return true;
        }
    }

    if (state.hasBlankLine) {
        if (node.children != null and node.children.?.len == 1) {
            const first_child = if (node.children) |children| blk: {
                if (children.len > 0) break :blk children[0] else break :blk null;
            } else null;

            if (first_child == null or first_child.?.children.?.len == 0) {
                return false;
            }
        }
    }

    if (isNewLine(char)) {
        return true;
    }

    const idx2 = state.openNodes.items.len;
    if (idx2 > 0) {
        const open_node = state.openNodes.items[idx2 - 1];
        if (std.mem.eql(u8, open_node.type, "paragraph")) {
            return true;
        }
    }

    const last_item = if (node.children) |children| blk: {
        if (children.len > 0) break :blk children[children.len - 1] else break :blk null;
    } else null;

    if (last_item) |li| {
        if (std.mem.eql(u8, li.type, "list_item") and state.indent >= li.subindent) {
            return true;
        }
    }

    return false;
}

pub fn isLooseList(node: *MarkdownNode) bool {
    var loose = false;

    if (node.children) |children| {
        if (children.len > 1) {
            var i: usize = 0;
            while (i < children.len - 1) : (i += 1) {
                const child = children[i];

                if (child.children) |cc| {
                    if (cc.len > 0) {
                        const grandchild = cc[cc.len - 1];
                        if (grandchild.blankAfter) {
                            child.blankAfter = true;
                        }
                    }
                }

                if (child.blankAfter) {
                    loose = true;
                    break;
                }
            }
        }

        for (children) |child| {
            if (child.children) |cc| {
                if (cc.len > 1) {
                    var j: usize = 0;
                    while (j < cc.len - 1) : (j += 1) {
                        const first = cc[j];
                        const second = cc[j + 1];
                        if (first.block and first.blankAfter and second.block) {
                            loose = true;
                            break;
                        }
                    }
                }
            }
            if (loose) break;
        }
    }

    return loose;
}

pub const ListInfo = struct {
    delimiter: u8,
    markup: []const u8,
    is_blank: bool,
    type: []const u8,
};
